tuf/ngclient/_internal/trusted_metadata_set.py

```diff
@@ -0,0 +1,464 @@
+# Copyright the TUF contributors
+# SPDX-License-Identifier: MIT OR Apache-2.0
+
+"""Trusted collection of client-side TUF Metadata
+
+TrustedMetadataSet keeps track of current valid set of metadata for the client,
+and handles almost every step of the "Detailed client workflow" (
+https://theupdateframework.github.io/specification/latest#detailed-client-workflow)
+in the TUF specification: the remaining steps are related to filesystem and
+network IO which is not handled here.
+
+Loaded metadata can be accessed via the index access with rolename as key
+(trusted_set["root"]) or, in the case of top-level metadata using the helper
+properties (trusted_set.root).
+
+The rules for top-level metadata are
+ * Metadata is loadable only if metadata it depends on is loaded
+ * Metadata is immutable if any metadata depending on it has been loaded
+ * Metadata must be loaded/updated in order:
+   root -> timestamp -> snapshot -> targets -> (other delegated targets)
+
+
+Exceptions are raised if metadata fails to load in any way.
+
+Example of loading root, timestamp and snapshot:
+
+>>> # Load local root (RepositoryErrors here stop the update)
+>>> with open(root_path, "rb") as f:
+>>>     trusted_set = TrustedMetadataSet(f.read())
+>>>
+>>> # update root from remote until no more are available
+>>> with download("root", trusted_set.root.signed.version + 1) as f:
+>>>     trusted_set.update_root(f.read())
+>>> # ...
+>>> trusted_set.root_update_finished()
+>>>
+>>> # load local timestamp, then update from remote
+>>> try:
+>>>     with open(timestamp_path, "rb") as f:
+>>>         trusted_set.update_timestamp(f.read())
+>>> except (RepositoryError, OSError):
+>>>     pass # failure to load a local file is ok
+>>>
```

Comment:
the spec is basically silent about loading anything local (except the trusted root), as far as I can tell: 
* it still expects us to compare new files to "trusted metadata" for rollback protection. I think loading a local copy is the logical way to find "trusted metadata".
* I think there's possibly even a case to be made for loading _expired_ local metadata for purposes of rollback prevention but since the spec explicitly mentions in the root case that expired metadata is fine, and does not mention that in other cases, I didn't seriously consider that. Still this may be a discussion topic ...

Reply:
I changed my mind on this while doing the review and should have deleted this comment.

> * it still expects us to compare new files to "trusted metadata" for rollback protection. I think loading a local copy is the logical way to find "trusted metadata".

Exactly this. The spec implies there's loading of snapshot and timestamp to check for rollback attacks and targets to check for freeze attacks. I made a note while reviewing this PR to do _something_ about this for the spec.

You have a good point that expired metadata is still valid as a source of information for protection against rollback attacks.